freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UsersyncPoller.java

```diff
@@ -113,13 +99,67 @@ void syncFreeIpaStacks() {
     }
 
     @VisibleForTesting
-    boolean isStale(Stack stack, UmsEventGenerationIds currentGeneration) {
+    boolean isStale(Stack stack, UserSyncStatus userSyncStatus, UmsEventGenerationIds currentGeneration) {
         try {
-            UserSyncStatus userSyncStatus = userSyncStatusService.getOrCreateForStack(stack);
-            return !currentGeneration.equals(userSyncStatus.getUmsEventGenerationIds().get(UmsEventGenerationIds.class));
+            boolean stale = userSyncStatus == null ||
+                    userSyncStatus.getUmsEventGenerationIds() == null;
+            if (!stale) {
+                try {
+                    UmsEventGenerationIds lastUmsEventGenerationIds = userSyncStatus.getUmsEventGenerationIds().get(UmsEventGenerationIds.class);
+                    stale = !currentGeneration.equals(lastUmsEventGenerationIds);
+                } catch (IOException e) {
+                    LOGGER.warn("Failed to retrieve UmsEventGenerationIds for Environment {} in Account {}. Assuming stale",
+                            stack.getEnvironmentCrn(), stack.getAccountId());
+                    stale = true;
+                }
+            }
+            LOGGER.debug("Environment {} in Account {} {} stale", stack.getEnvironmentCrn(), stack.getAccountId(), stale ? "is" : "is not");
+            return stale;
         } catch (Exception e) {
             LOGGER.warn("Unable to calculate staleness due to exception.", e);
             return false;
         }
     }
+
+    @VisibleForTesting
+    boolean isCooldownExpired(Stack stack, UserSyncStatus userSyncStatus, long cooldownThresholdTime) {
+        try {
+            boolean cool = userSyncStatus == null ||
+                    userSyncStatus.getLastFullSyncStartTime() == null ||
+                    userSyncStatus.getLastFullSyncStartTime() < cooldownThresholdTime;
+            LOGGER.debug("Environment {} in Account {} {} cool", stack.getEnvironmentCrn(), stack.getAccountId(), cool ? "is" : "is not");
+            return cool;
+        } catch (Exception e) {
+            LOGGER.warn("Unable to determine if cooldown expired due to exception.", e);
+            return false;
+        }
```

Comment:
This is weird. There is nothing that can really throw here, no? Maybe remove.

Reply:
removed